core/server/src/main/java/alluxio/master/journal/JournalWriter.java

```diff
@@ -76,12 +84,37 @@
     mJournal = Preconditions.checkNotNull(journal);
     mJournalDirectory = mJournal.getDirectory();
     mCompletedDirectory = mJournal.getCompletedDirectory();
+    mCheckpointPath = mJournal.getCheckpointFilePath();
     mTempCheckpointPath = mJournal.getCheckpointFilePath() + ".tmp";
+    mTempBackupCheckpointPath = mJournal.getCheckpointFilePath() + ".backup.tmp";
+    mBackupCheckpointPath = mJournal.getCheckpointFilePath() + ".backup";
     mUfs = UnderFileSystem.get(mJournalDirectory);
     mMaxLogSize = Configuration.getBytes(PropertyKey.MASTER_JOURNAL_LOG_SIZE_BYTES_MAX);
   }
 
   /**
+   * Recovers the checkpoint file in case the master crashed while updating it previously.
+   */
+  public synchronized void recoverCheckpoint() {
+    try {
+      if (mUfs.exists(mBackupCheckpointPath)) {
+        if (mUfs.exists(mJournal.getCheckpointFilePath())) {
+          // We must have crashed while cleaning up the completed logs directory and backup
+          // checkpoint, so we finish these steps now.
+          deleteCompletedLogs();
+          mUfs.delete(mBackupCheckpointPath, false);
+        } else {
+          // We must have crashed before writing the checkpoint file, restore the checkpoint from
```

Comment:
I'm not sure when this would happen, so if there is a javadoc for this class that describes the checkpointing process, that would be helpful.